bind-contrib/contrib/updatehosts-1.1.0/man/updatehosts.1
.\" $Id: updatehosts.1,v 1.1 2000/07/11 06:53:22 vixie Exp $
.TH UPDATEHOSTS 1 $Date: 2000/07/11 06:53:22 $
.UC 4
.SH NAME
updatehosts \- update DNS and static host tables
.SH SYNOPSIS
.B updatehosts
[
.B -d dbdirectory
]
[
.B -f configfile
]
[
.B -l
]
[
.B -n
]
[
.B -s
]
[
.B -v
]
.B [ file ... ]
.SH DESCRIPTION
Updates the DNS and static host database zone files.
.PP
.B updatehosts
is a front-end to the
.B gendns
and
.B genstatic
programs which generate DNS zone files the and static host file from a set
of input files organized as a set of flat relational files.
.B updatehosts
is driven off of a configuration file usually found in
.B /usr/local/etc/updatehosts.env
which sets variables such as the location of the input
database files.
An alternate configuration file location can be specified by the -f option.
.PP
The input database files are normally kept in the directory
.B /var/named/db.
The location of this directory can be specified by the -d option or with the
DB_DIR environment variable.
Output from
.B updatehosts
is normally placed in the
.B /var/named
directory.
An alternate output directory location may be specified by setting the
NAMED_DIR variable.
.PP
The following files are maintained in the DB_DIR directory:
.TP
main
Contains information about each host within a set of domains.  Information
such as the domain name of each host and its IP address are maintained in
this file.
.TP
cname
Contains information about aliases for hosts
.TP
global
Contains some global parameters
.TP
mx
Contains information about MX (mail exchanger) records.
.TP
options
Contains information about BIND V8 optios.
.TP
ns
Contains information about NS (nameserver) records.
.TP
secondary
Contains information about secondary zones served by this server
.TP
soa
Contains information about SOA (start of authority) records.
.TP
txt
Contains information about TXT (text) records.
.TP
subnet
Contains information about local subnets which is used to generate an
optional BOOTP or DHCP configuration file.
.PP
Each of the database files is organized tabularly.  These files are
preprocessed by the
.B readinfo
command.
(see readinfo.1)
.PP
.B updatehosts
may takes arguments which are the names of the database files which the user
wants to edit.  All database files may optionally kept under SCCS control
by setting the USE_SCCS variable or by placing the -s flag on
the command line.
Similarly RCS control can be maintained by setting the USE_RCS
variable or setting the -r flag on the command line.
A user using
.B updatehosts
must have write permission in the DB_DIR and the NAMED_DIR directory.
Each database file is edited by the
.B vi
editor or whatever program the user has set in the EDITOR environment
variable.
.PP
After each edit, the database files are optionally checked back in using SCCS
or RCS.
Then the static host file is generated using the
.B genstatic
command.
The host file is normally placed in the NAMED_DIR directory under the name
.B static.
This name may be changed by setting the STATICHOSTS variable.
.PP
Then the DNS zone files and
bootstrap file are updated using the
.B gendns
command.
Both BIND Version 4 or Version 8 configuration files are generated.
.PP
The local nameserver is restarted
with the newly updated zone and bootstrap files by calling
.B "poke_ns restart"
or
.B poke_ns reload
if the -l flag is specified
.PP
Optionally, the POSTPROCESS environment variable can be set to the name of
a command to run after the nameserver zone files and static host files are
generated.
This can be any arbitrary command.
Normally, this may be a command to update NIS or NIS+ host maps from the
generated static host file.
.PP
Environment variables should normally be set in the configuration file
except when testing.  An exception to this rule is the EDITOR environment
variable which is normally set to the user's favorite editor.
.SH
Normal Configuration Setup
.PP
.B updatehosts
provides a great deal of flexibility in how a nameserver zone files and
static host table are generated and the various permissions on those files.
In normal practice it is desirable to set up a group (say, staff) to manage
the nameserver files.
Setting the directory permissions on the NAMED_DIR tree to allow staff to
create and modify files in that tree allows
.B updatehosts
to be run by anyone in the staff group without root access.
The
.B poke_ns
program must be run as root to restart the nameserver.
So setting it setuid to root, but only executable by the staff group allows
any member of the staff group to restart the running nameserver. 
.SH
See also
.PP
.BR gendns(1),
.BR genstatic(1),
.BR poke_ns(1)


CHANGES
5500.	[bug]		Fake entry.

5494.	[bug]		Silence the EPROTO syslog message on older systems.
			[GL #1928]

5493.	[bug]		Fix off-by-one error when calculating new hashtable
			size. [GL #2104]

5492.	[bug]		Tighten LOC parsing to reject period and/or m as a
			value.  Correct handling of negative altitudes which
			are not whole metres.  [GL #2074]

5491.	[bug]		rbtversion->glue_table_size could be read without the
			appropriate lock being held. [GL #2080]

5490.	[func]		Refactor the readline support to use pkg-config and
			add support for editline library. [GL !3942]

5489.	[bug]		Named failed to reject some invalid records resulting
			in records that, after being printed, could not be
			loaded or would result in DNSSEC validation failures
			when re-read from zone files as the wire format
			differed.  The covered records records are: CERT,
			IPSECKEY, NSEC3, NSEC3PARAM, NXT, SIG, TLSA, WKS, and
			X25. [GL !3953]

5488.	[bug]		nta needed to have a weak reference on view to prevent
			the view being deleted while nta tests are being
			performed. [GL #2067]

5487.	[cleanup]	Update managed keys log messages to be less confusing.
			[GL #2027]

5486.	[func]		Add 'rndc dnssec -checkds' command to tell named
			that the DS record has been published in the parent.
			[GL #1613]

	--- 9.17.4 released ---

5485.	[placeholder]

5484.	[func]		Expire zero TTL records quickly rather than using them
			for stale answers. [GL #1829]

5483.	[func]		Keeping "stale" answers in cache has been disabled by
			default and can be re-enabled with a new configuration
			option "stale-cache-enable". [GL #1712]

5482.	[bug]		If the Duplicate Address Detection (DAD) mechanism had
			not yet finished after adding a new IPv6 address to the
			system, BIND 9 would fail to bind to IPv6 addresses in a
			tentative state. [GL #2038]

5481.	[security]	"update-policy" rules of type "subdomain" were
			incorrectly treated as "zonesub" rules, which allowed
			keys used in "subdomain" rules to update names outside
			of the specified subdomains. The problem was fixed by
			making sure "subdomain" rules are again processed as
			described in the ARM. (CVE-2020-8624) [GL #2055]

5480.	[security]	When BIND 9 was compiled with native PKCS#11 support, it
			was possible to trigger an assertion failure in code
			determining the number of bits in the PKCS#11 RSA public
			key with a specially crafted packet. (CVE-2020-8623)
			[GL #2037]

5479.	[security]	named could crash in certain query resolution scenarios
			where QNAME minimization and forwarding were both
			enabled. (CVE-2020-8621) [GL #1997]

5478.	[security]	It was possible to trigger an assertion failure by
			sending a specially crafted large TCP DNS message.
			(CVE-2020-8620) [GL #1996]

5477.	[bug]		The idle timeout for connected TCP sockets, which was
			previously set to a high fixed value, is now derived
			from the client query processing timeout configured for
			a resolver. [GL #2024]

5476.	[security]	It was possible to trigger an assertion failure when
			verifying the response to a TSIG-signed request.
			(CVE-2020-8622) [GL #2028]

5475.	[bug]		Wildcard RPZ passthru rules could incorrectly be
			overridden by other rules that were loaded from RPZ
			zones which appeared later in the "response-policy"
			statement. This has been fixed. [GL #1619]

5474.	[bug]		dns_rdata_hip_next() failed to return ISC_R_NOMORE
			when it should have. [GL !3880]

5473.	[func]		The RBT hash table implementation has been changed
			to use a faster hash function (HalfSipHash2-4) and
			Fibonacci hashing for better distribution. Setting
			"max-cache-size" now preallocates a fixed-size hash
			table so that rehashing does not cause resolution
			brownouts while the hash table is grown. [GL #1775]

5472.	[func]		The statistics channel has been updated to use the
			new network manager. [GL #2022]

5471.	[bug]		The introduction of KASP support inadvertently caused
			the second field of "sig-validity-interval" to always be
			calculated in hours, even in cases when it should have
			been calculated in days. This has been fixed. (Thanks to
			Tony Finch.) [GL !3735]

5470.	[port]		gsskrb5_register_acceptor_identity() is now only called
			if gssapi_krb5.h is present. [GL #1995]

5469.	[port]		On illumos, a constant called SEC is already defined in
			<sys/time.h>, which conflicts with an identically named
			constant in libbind9. This conflict has been resolved.
			[GL #1993]

5468.	[bug]		Addressed potential double unlock in process_fd().
			[GL #2005]

5467.	[func]		The control channel and the rndc utility have been
			updated to use the new network manager. To support
			this, the network manager was updated to enable
			the initiation of client TCP connections. Its
			internal reference counting has been refactored.

			Note: As a side effect of this change, rndc cannot
			currently be used with UNIX-domain sockets, and its
			default timeout has changed from 60 seconds to 30.
			These will be addressed in a future release.
			[GL #1759]

5466.	[bug]		Addressed an error in recursive clients stats reporting.
			[GL #1719]

5465.	[func]		Added fallback to built-in trust-anchors, managed-keys,
			or trusted-keys if the bindkeys-file (bind.keys) cannot
			be parsed. [GL #1235]

5464.	[bug]		Requesting more than 128 files to be saved when rolling
			dnstap log files caused a buffer overflow. This has been
			fixed. [GL #1989]

5463.	[placeholder]

5462.	[bug]		Move LMDB locking from LMDB itself to named. [GL #1976]

5461.	[bug]		The STALE rdataset header attribute was updated while
			the write lock was not being held, leading to incorrect
			statistics. The header attributes are now converted to
			use atomic operations. [GL #1475]

5460.	[cleanup]	tsig-keygen was previously an alias for
			ddns-confgen and was documented in the ddns-confgen
			man page. This has been reversed; tsig-keygen is
			now the primary name. [GL #1998]


bind-contrib/contrib/updatehosts-1.1.0/updatehosts.sh
#! @SH@
# $Id: updatehosts.sh,v 1.1 2000/07/11 06:53:12 vixie Exp $

# updatehosts DNS maintenance package
# Copyright (C) 1998-2000  Smoot Carl-Mitchell
# 
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
# 
# smoot@tic.com

# update the host tables and DNS files
#
# arguments are the files to edit

CONFIG_FILE=@sysconfdir@/updatehosts.env
DB_DIR=/var/named/db
EDITOR=${EDITOR-vi}
NAMED_DIR=/var/named
VERSION=8
RELOAD=1
RESTART=

# check if programs exists in user's PATH
in_path () {
	for prog do
		which $prog 2>&1 >/dev/null || return 1
	done
	return 0
}

# look for the startup file which sets global environment variables
[ -f $CONFIG_FILE ] && . $CONFIG_FILE

# read the options - command line overrides the configuration file
while getopts FVd:f:lnsv option; do
	case $option in
	F) FORCE_ZONE=-F;;
	d) DB_DIR=$OPTARG ;;
	f) CONFIG_FILE=$OPTARG
	   [ -f $CONFIG_FILE ] && . $CONFIG_FILE;;
	l) RELOAD=1 ;;
	n) RESTART=1 ;;
	r) USERCS=1 ;;
	s) USESCCS=1 ;;
	v) VERBOSE=-v ;;
	V) PR_VERSION=1 ;;
	esac
done
[ $OPTIND ] && shift `expr $OPTIND - 1`

if [ $PR_VERSION ]; then
	echo Updatehosts Version 1.1.0
	exit 0;
fi
# set the overridden variables
# unset Boolean variables set to "0" or "no"
[ "$GENBOOTP" = "0" -o "$GENBOOTP" = "no" ] && GENBOOTP=
[ "$GENDHCP" = "0" -o "$GENDHCP" = "no" ] && GENDHCP=
[ "$USERCS" = "0" -o "$USERCS" = "no" ] && USERCS=
[ "$USESCCS" = "0" -o "$USESCCS" = "no" ] && USESCCS=

list=
# check the filename arguments
for i do
	found=
	for j in $DB_FILES; do
		if [ $i = $j ]; then
			found=1
			break
		fi
	done
	if [ ! "$found" ]; then
		echo $i is not a valid filename
		list=1
	fi
done
if [ "$list" ]; then
	echo legal filenames are:
	echo $DB_FILES
	exit 1
fi

cd $NAMED_DIR || ( echo no $NAMED_DIR directory; exit 1 )

# set umask to group read and write
umask 2
# check if the files exist and are under SCCS control and are readable
#
# files are in a subdirectory - so we cd there
cd $DB_DIR || ( echo no $DB_DIR present; exit 1 )

okay=1
if [ "$USESCCS" ]; then
	# check for presence of sccs command
	if in_path sccs; then
		:
	else
		echo "cannot find sccs command"
		okay=0
	fi
	for i do
		if [ ! -f SCCS/s.$i ]; then
			echo "file $i is not under SCCS control"
			okay=0
		elif [ ! -r SCCS/s.$i ]; then
			echo "file $i is not readable by you"
			okay=0
		elif [ -f SCCS/p.$i ]; then
			echo "file $i is already checked out under SCCS"
			okay=0
		elif [ -f $i ]; then
			rm -f $i
		fi
	done
elif [ "$USERCS" ]; then
	# check for presence of rcs command
	if in_path co ci; then
		:
	else
		echo "cannot find rcs co or ci command"
		okay=0
	fi
	for i do
		if [ ! -f RCS/$i,v ]; then
			echo "file $i is not under RCS control"
			okay=0
		elif [ ! -r RCS/$i,v ]; then
			echo "file $i is not readable by you"
			okay=0
		elif [ -f $i ]; then
			rm -f $i
		fi
	done
fi

if [ $okay -eq 0 ]; then
	exit 1
fi

# checkout and edit the files
for i do
	if [ "$USESCCS" ]; then
		sccs edit $i
	elif [ "$USERCS" ]; then
		co -l $i
	fi
	$EDITOR $i
	if [ "$USESCCS" ]; then
		sccs delget $i
	elif [ "$USERCS" ]; then
		ci -u $i
	fi
done

cd $NAMED_DIR

# generate the static tables
if [ $STATICHOSTS ]; then
	echo "generating static hosts table..."
	HOSTS=$STATICHOSTS genstatic
fi

# generate the dns zone files
echo "generating the DNS database..."
[ ! "$ERROR_LEVEL" ] && ERROR_LEVEL=2
gendns -e $ERROR_LEVEL $FORCE_ZONE $VERBOSE -d $DB_DIR

# generate the bootptab file 
# but only if we are instructed to
if [ $GENBOOTP ]; then
	echo "generating the bootptab file..."
	genbootp

	# copy it to alternate machines
	for m in $SECBOOTP; do
		cp -p bootptab /net/$SECBOOTP/etc
	done
fi

# generate the dhcptab file 
# but only if we are instructed to
if [ $GENDHCP ]; then
	echo "generating the dhcpd.conf file..."
	if [ $VERBOSE ]; then
		VERBOSE=-v
	fi
	if [ $KNOWNHOSTS ]; then
		known=-k
	fi
	gendhcp $VERBOSE $known

	# copy it to alternate machines
	for m in $SECDHCP; do
		cp -p dhcpd.conf /net/$SECDHCP/etc/dhcpd.conf
	done

	# restart the server
	echo "restarting DHCP server..."
	poke_dhcp >/dev/null 2>&1
fi

# run postprocessor command
if [ "$POSTPROCESS" ]; then
	echo running $POSTPROCESS
	eval $POSTPROCESS
fi

if [ $RELOAD ]; then
	echo "reloading zone files..."
	poke_ns reload
elif [ $RESTART ]; then
	echo "restarting nameserver..."
	poke_ns restart
fi
